### PR TITLE
hyprland: 0.40.0-unstable-2024-05-05 -> 0.41.1; hyprwayland-scanner: 0.3.4 -> 0.3.10; hyprlandPlugins.hy3: 0.40.0 -> 0.41.0; hyprutils: init at 0.1.2 

### LINKED ISSUE #317234: hyprland: hyprctl version missing
### Describe the bug
hyprland is no longer able to present its version. Some other tools depend on getting this version for parsing hyprctl outputs.
[This commit](https://github.com/NixOS/nixpkgs/commit/23d1b8e4fececf2c83875812c0920c3ecd933976#diff-f466618ed484517008b9ba6835831b232aced0d90cd7499d31b819dcb7cf99b1L67-L75) removed the inserted version info.

If/When solving this issue there could also be a problem with a newly introduced placeholder as mentioned in [this comment](https://github.com/NixOS/nixpkgs/pull/309075#discussion_r1593987493).

### Steps To Reproduce
`hyprctl version` or `hyprctl version -j` does not show the current version of hyprland.

### Expected behavior
`hyprctl version` or `hyprctl version -j` shows the current version of hyprland.

### Screenshots
image

### Notify maintainers
@wozeparrot @fufexan 

### Metadata
```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.9.3, NixOS, 24.05 (Uakari), 24.05.803.b3b2b28c1daa`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - channels(root): `"nixos-24.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```
---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by JohnRTitor**
## Description of changes

- Update `Hyprland` to v0.41.1. Changelog: [1](https://github.com/hyprwm/Hyprland/releases/tag/v0.41.0), [2](https://github.com/hyprwm/Hyprland/releases/tag/v0.41.1)
- Add missing deps
- Update `hyprwayland-scanner` too since v0.3.10 is required by Hyprland 0.41.0
- `hyprlandPlugins.hy3`: 0.40.0 -> 0.41.0
- Fixes #317234

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
---

CC: @fufexan @wineee 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by fufexan — COMMENTED** [review]
Can you also update hyprlandPackages.hy3?

**Comment by JohnRTitor** [comment]
@fufexan can hyprland-protocols be safely removed since v40.0 does not use it anymore?

**Review by aacebedo — APPROVED** [review]

**Comment by leiserfg** [comment]
@fufexan given that 0.40.0 has some issues with external monitors, can this be backported to the stable channel?

File: pkgs/applications/window-managers/hyprwm/hyprland/default.nix [hunk 1]
```diff
@@ -53,41 +56,46 @@ assert lib.assertMsg (!hidpiXWayland) "The option `hidpiXWayland` has been remov
...
   postPatch = ''
     # Fix hardcoded paths to /usr installation
     sed -i "s#/usr#$out#" src/render/OpenGL.cpp
+
+    # Remove extra @PREFIX@ to fix pkg-config paths
+    sed -i "s#@PREFIX@/##g" hyprland.pc.in
   '';
 
   # used by version.sh
-  DATE = "2024-05-05";
```
Comment (fufexan) [inline pkgs/applications/window-managers/hyprwm/hyprland/default.nix:72]: Why is this removed? It's still being picked up by `version.sh`.

**Comment by fufexan** [comment]
Fixes https://github.com/NixOS/nixpkgs/issues/317234.

File: pkgs/applications/window-managers/hyprwm/hyprland/default.nix (hunk 1 above)
Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:72]: I thought that was for just unstable git Hyprland?

Is there a way to do automate it, perhaps on the source side, a file to fetch which contains the date?

Comment (fufexan) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:72]: Technically git should provide the date, as it does when using flakes. But I don't see it being provided by `fetchgit`.

Comment (leiserfg) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:72]: I think that requires `leaveDotGit = true;` but it turns the output nondeterministic.

Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:72]: Another thing, we can't just use this anymore, since we are moving to a tagged release.
https://github.com/NixOS/nixpkgs/blob/07c3ec6a8975b940e96f1db59d5650a3b25bf73a/pkgs/applications/window-managers/hyprwm/hyprland/default.nix#L78

Comment (leiserfg) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:72]: No, but you get the `.git` that could be used to extract the date. But again it turns the output nondeterministic as .git will change with upstream 🤔

Comment (leiserfg) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:72]: Could this be used? 
https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/set-source-date-epoch-to-latest.sh

Comment (fufexan) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:72]: > Another thing, we can't just use this anymore, since we are moving to a tagged release.
> 
> https://github.com/NixOS/nixpkgs/blob/07c3ec6a8975b940e96f1db59d5650a3b25bf73a/pkgs/applications/window-managers/hyprwm/hyprland/default.nix#L78

Right, let's drop it then. We can drop the date as well, if we only build tagged releases.

Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:72]: Done in https://github.com/NixOS/nixpkgs/pull/319131/commits/f6b173654a0b22be3087e97d622f3b46603f92fd

**Comment by fufexan** [comment]
@JohnRTitor I just remembered, we still need to provide *some* amount of info for `hyprctl version`, otherwise we can't fix #317234. We can set `TAG = finalAttrs.src.rev;` and `BRANCH = "main";`.
Though I'd still like to provide both the date and the commit somehow, even if they're manually populated. Otherwise I can cook up an update script to get those.

**Comment by JohnRTitor** [comment]
Yeah, an update script would be nice to have! 
Something like [this](https://github.com/JohnRTitor/nixpkgs/blob/20362ae38f86c948b5505d7a006474804cf038ee/pkgs/tools/filesystems/bcachefs-tools/default.nix#L107-L115) should work, as a base. 

Post the updotu

**Comment by fufexan** [comment]
[hyprland_updater.patch.txt](https://github.com/user-attachments/files/15810542/hyprland_updater.patch.txt)
Might be a pretty crude approach, and it's definitely ugly, but it does the job well.

`hyprctl version` output:
```
› hyprctl version              
Hyprland, built from branch main at commit ea2501d4556f84d3de86a4ae2f4b22a474555b9f  (props: bump version to 0.41.0).
Date: 2024-06-10
Tag: v0.41.0, commits: ea2501d4556f84d3de86a4ae2f4b22a474555b9f

flags: (if any)
```

**Comment by JohnRTitor** [comment]
> P.S.: while you're at it, try seeing if the script correctly updates to 0.41.1.

It does!

**Review by fufexan — CHANGES_REQUESTED** [review]
Please also squash the commits per-package.

File: pkgs/applications/window-managers/hyprwm/hyprland/default.nix [hunk 2]
```diff
@@ -1,93 +1,111 @@
...
-    hwdata
-    hyprwayland-scanner
-    jq
-    makeWrapper
-    cmake
-    meson # for wlroots
-    ninja
-    pkg-config
-    wayland-scanner
-    python3 # for udis86
```
[the remark this answers is not in the document]
Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:90]:
```
    wlroots.buildInputs
    udis86.buildInputs
```

Is it safe to remove these then?

**Review by donovanglover — COMMENTED** [review]
v0.41.1 has a new dependency [hyprutils](https://github.com/hyprwm/hyprutils). Build fails as-is right now

Edit: Got it to work with [this package](https://github.com/NixOS/nixpkgs/commit/5c0406eed2d24f6d28138c6ce1bbb4195a4cd003) if you want to copy it. Make sure to add `meta.maintainers` and add the package to `buildInputs`.

File: pkgs/applications/window-managers/hyprwm/hyprland/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (fufexan) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix:90]: Yes, those can be removed.

[the remark this answers is not in the document]
Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix]: Done. Builds fine now.

**Comment by JohnRTitor** [comment]
> Please also squash the commits per-package.

@fufexan Squashed a few, significant commits like adding a whole update script can stay in a separate commit. Please check.

File: pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix
Comment (fufexan) [inline pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix:1]: Was this moved to aid with the automatic `pkgs/by-name` migration?

Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix:1]: Well since I moved Hyprland package out  to by-name already, it made sense to move/rename the plugins file appropriately. 

The plugins directory is called recursively in `all-packages`, so I am not sure if it's possible to automatically migrate it without changing the structure. 

https://github.com/NixOS/nixpkgs/blob/2d88a5f17a118bbb3e4663196931b98df3321c81/pkgs/top-level/all-packages.nix#L5512

Comment (fufexan) [reply pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix:1]: It can be migrated and the entry in `all-packages` can be left as-is, while pointing to the new location.

Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix:1]: True, but isn't the purpose of `pkgs/by-name` is that we don't have to put an entry in `all-packages`? @Mic92 would love some input here.

Comment (Mic92) [reply pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix:1]: In theory we could move to `by-name/hy/hyperland/plugins.nix` as well so it lives in the same directory as hyperland itself. We would still need an entry for the plugins in all-packages.nix though, because these nested attrsets are not compatible with by-name.

Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix:1]: Yeah, that's what fufexan was referring to. To be honest I am more inclined on keeping the plugins here and `hyprland` itself in by-name.

Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix:1]: @fufexan is this blocked? If not I'd like to see it merged. 

Migrating the plugins to `by-name` the "actual" way would not be feasible due to the nature of the plugins derivation. 

Since hyprutils is not on Nixpkgs yet, I can't use `nixpkgs.follows` for Hyprland flake input!

Comment (fufexan) [reply pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix:1]: I'm fine with leaving it as is then.

**Review by fufexan — APPROVED** [review]
LGTM!

File: pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix
Comment (JohnRTitor) [reply pkgs/applications/window-managers/hyprwm/hyprland-plugins/default.nix:1]: Sweet, thanks

**Review by donovanglover — APPROVED** [review]
LGTM thanks!

**Comment by JohnRTitor** [comment]
Looks like the automatic backport failed, I am currently out of the city so I can not attend it.
Would be nice if anyone is kind enough to cherry-pick and backport these changes!
